<center><em>Copyright Pierian Data</em></center>
<center><em>For more information, visit us at <a href='http://www.pieriandata.com'>www.pieriandata.com</a></em></center>

# NumPy Operations

## Arithmetic

You can easily perform *array with array* arithmetic, or *scalar with array* arithmetic. Let's see some examples:

In [1]:
import numpy as np

In [2]:
arr = np.arange(0, 10)
arr

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [3]:
# for array, it happens on element by element
arr + 5

array([ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

In [4]:
arr + 2

array([ 2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [5]:
arr + arr

array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18])

In [6]:
arr - arr

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [7]:
arr * arr

array([ 0,  1,  4,  9, 16, 25, 36, 49, 64, 81])

In [8]:
# This will raise a Warning on division by zero, but not an error!
# It just fills the spot with nan
arr / arr

C:\Users\Phone Thiri Yadana\.conda\envs\venv-rf\lib\site-packages\ipykernel_launcher.py:1: RuntimeWarning: invalid value encountered in true_divide
  """Entry point for launching an IPython kernel.


array([nan,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.])

In [10]:
1 / 0

ZeroDivisionError: division by zero

In [9]:
# Also a warning (but not an error) relating to infinity
1 / arr

C:\Users\Phone Thiri Yadana\.conda\envs\venv-rf\lib\site-packages\ipykernel_launcher.py:2: RuntimeWarning: divide by zero encountered in true_divide
  


array([       inf, 1.        , 0.5       , 0.33333333, 0.25      ,
       0.2       , 0.16666667, 0.14285714, 0.125     , 0.11111111])

In [11]:
arr * 3

array([ 0,  3,  6,  9, 12, 15, 18, 21, 24, 27])

## Universal Array Functions

NumPy comes with many [universal array functions](http://docs.scipy.org/doc/numpy/reference/ufuncs.html), or <em>ufuncs</em>, which are essentially just mathematical operations that can be applied across the array.<br>Let's show some common ones:

In [12]:
# Taking Square Roots
np.sqrt(arr)

array([0.        , 1.        , 1.41421356, 1.73205081, 2.        ,
       2.23606798, 2.44948974, 2.64575131, 2.82842712, 3.        ])

In [15]:
# Calculating exponential (e^)
np.exp(arr)

array([1.00000000e+00, 2.71828183e+00, 7.38905610e+00, 2.00855369e+01,
       5.45981500e+01, 1.48413159e+02, 4.03428793e+02, 1.09663316e+03,
       2.98095799e+03, 8.10308393e+03])

In [13]:
# Trigonometric Functions like sine
np.sin(arr)

array([ 0.        ,  0.84147098,  0.90929743,  0.14112001, -0.7568025 ,
       -0.95892427, -0.2794155 ,  0.6569866 ,  0.98935825,  0.41211849])

In [14]:
# Taking the Natural Logarithm
np.log(arr)

C:\Users\Phone Thiri Yadana\.conda\envs\venv-rf\lib\site-packages\ipykernel_launcher.py:2: RuntimeWarning: divide by zero encountered in log
  


array([      -inf, 0.        , 0.69314718, 1.09861229, 1.38629436,
       1.60943791, 1.79175947, 1.94591015, 2.07944154, 2.19722458])

## Summary Statistics on Arrays

NumPy also offers common summary statistics like <em>sum</em>, <em>mean</em> and <em>max</em>. You would call these as methods on an array.

In [17]:
arr

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [18]:
np.sum(arr)

45

In [19]:
np.mean(arr)

4.5

In [20]:
np.min(arr)

0

In [22]:
np.max(arr)

9

In [23]:
np.var(arr)

8.25

In [24]:
np.std(arr)

2.8722813232690143

<strong>Other summary statistics include:</strong>
<pre>
arr.min() returns 0                   minimum
arr.var() returns 8.25                variance
arr.std() returns 2.8722813232690143  standard deviation
</pre>

## Axis Logic
When working with 2-dimensional arrays (matrices) we have to consider rows and columns. This becomes very important when we get to the section on pandas. In array terms, axis 0 (zero) is the vertical axis (rows), and axis 1 is the horizonal axis (columns). These values (0,1) correspond to the order in which <tt>arr.shape</tt> values are returned.

Let's see how this affects our summary statistic calculations from above.

In [25]:
arr_2d = np.arange(0, 25).reshape(5,5)

In [26]:
arr_2d

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24]])

In [27]:
arr_2d.sum()

300

In [29]:
arr_2d.sum(axis=0) # across rows

array([50, 55, 60, 65, 70])

By passing in <tt>axis=0</tt>, we're returning an array of sums along the vertical axis, essentially <tt>[(1+5+9), (2+6+10), (3+7+11), (4+8+12)]</tt>

<img src='axis_logic.png' width=400/>
<img src='axis_logic1.png' width=400/>

In <tt>arr_2d.sum(axis=0)</tt> above, the first element in each row was summed, then the second element, and so forth.

So what should <tt>arr_2d.sum(axis=1)</tt> return?

In [30]:
arr_2d.sum(axis=1) # across columns

array([ 10,  35,  60,  85, 110])


<img src='axis_logic2.png' width=400/>

# Great Job!

That's all we need to know for now!